metabase/metabase
# Add custom warning message when leaving changing Data Model (segments and metrics)

Closes #33795

### Description

The PR is both for segments & metrics because those components are nearly identical. The changes should look almost the same in both:
- `MetricForm.tsx` & `SegmentForm.tsx`
- `MetricApp.jsx` & `SegmentApp..jsx`
- `MetricApp.unit.spec.tsx` & `SegmentApp.unit.spec.tsx`
- `metric.ts` & `segment.ts`

----

- Renamed `MetricForm.unit.spec.tsx` & `SegmentForm.unit.spec.tsx` to `MetricApp.unit.spec.tsx` & `SegmentApp.unit.spec.tsx` because those are the components that these specs are testing
- Moved `MetricApp.unit.spec.tsx` & `SegmentApp.unit.spec.tsx` to near the components that these specs are testing - unfortunately GitHub does not detect it as a move, but as file deletion + addition
- Made `MetricApp.unit.spec.tsx` & `SegmentApp.unit.spec.tsx` look nearly identical - imports, setup, assertions, etc.
- Improved a11y of `MetricForm` & `SegmentForm` by adding `htmlFor` and `id` attributes
  - now users can click on a form label to focus form input
  - now we can use `getByLabelText` instead of `getByPlaceholderText` (placeholders were really long there)
  - this will obviously improve compatibility with screen readers
- Converted `UpdateSegmentForm`, `UpdateSegmentFormInner`, `CreateSegmentForm` & `CreateSegmentFormInner`, `MetricApp` & `SegmentApp` to functional components ~so that `useCallbackEffect` can be used in them~
- Similar to:
  - #34084 
  - #34192
  - #33827
  - #34272
  - #34316 

### How to verify

Scenarios are identical for metrics (`admin/datamodel/metrics`) and segments (`admin/datamodel/segments`).
I'll list scenarios for metrics only.

Prerequisite for all scenarios:

1. Open Metrics page: http://localhost:3000/admin/datamodel/metrics

**1. Custom warning modal is shown when leaving creating a metric with unsaved changes**

1. Click "New Metric"
2. Change something in the metric definition
3. Navigate away from the page (browser back history button, "Cancel" button, link in page header, etc.)

Expected result: custom warning modal **is shown**.

**2. Custom warning modal is not shown when creating a metric**

1. Click "New Metric"
2. Fill out the form
3. Click "Save changes"

Expected result: custom warning modal **is not shown**, you land on Metrics page.

**3. Custom warning modal is shown when leaving editing a metric with unsaved changes**

1. Edit metric created in scenario 2
2. Change something in the form
3. Navigate away from the page (browser back history button, "Cancel" button, link in page header, etc.)

Expected result: custom warning modal **is shown**.

**4. Custom warning modal is not shown when leaving creating a metric with unsaved changes**

1. Click "New Metric"
2. Navigate away from the page (browser back history button, "Cancel" button, link in page header, etc.)

Expected result: custom warning modal **is not shown**.

**5. Custom warning modal is not shown when leaving creating a metric with unsaved changes**

1. Edit metric created in scenario 2
2. Navigate away from the page (browser back history button, "Cancel" button, link in page header, etc.)

Expected result: custom warning modal **is not shown**.

### Demo

https://github.com/metabase/metabase/assets/6830683/7415c31e-2bac-4ff0-adc2-bf739942a264
